### PR TITLE
Kotlin-Server: Optional metrics and upgrade to Ktor 1.5.4

### LINKED ISSUE #828: [kotlin-server] Paths.kt does not contain post-operations
##### Description

When generating a kotlin-server, the generated Paths.kt does not contain locations for post-operations.

##### Swagger-codegen version

2.3.1

##### Swagger declaration file content or url

```yaml
swagger: "2.0"
host: "0.0.0.0"
info:
  title: "Kotlin Demo"
  description: "This is a Kotlin Swagger Demo"
  version: "0.0.1"
basePath: "/v1"
paths:
  /user/{userId}:
    get:
      operationId: "getUser"
      produces:
      - "application/json"
      responses:
        200:
          description: "Success."
          schema:
            $ref: '#/definitions/IdentifiableUser'
      parameters:
      - name: "userId"
        in: "path"
        required: true
        type: "integer"
  /user:
    post:
      operationId: "postUser"
      produces:
      - "application/json"
      parameters:
      - in: "body"
        name: "body"
        required: true
        schema:
          $ref: "#/definitions/User"
      responses:
        202:
          description: "Success"
        400:
          description: "Invalid User."
definitions:
  Identifiable:
    type: "object"
    required:
    - "id"
    properties:
      id:
        type: "integer"
  User:
    type: "object"
    required:
    - "name"
    - "type"
    properties:
      name:
        type: "string"
      type:
        type: "string"
        enum: ["user", "moderator"]
      info:
        type: "string"
  IdentifiableUser:
    type: "object"
    allOf:
    - $ref: "#/definitions/Identifiable"
    - $ref: "#/definitions/User"
```

Results in:

```kotlin
/**
* Kotlin Demo
* This is a Kotlin Swagger Demo
*
* OpenAPI spec version: 0.0.1
* 
*
* NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
* https://openapi-generator.tech
* Do not edit the class manually.
*/
package org.openapitools.server

import io.ktor.application.ApplicationCall
import io.ktor.http.HttpMethod
import io.ktor.locations.*
import io.ktor.pipeline.PipelineContext
import io.ktor.routing.Route
import io.ktor.routing.method
import org.openapitools.server.models.*


// NOTE: ktor-location@0.9.0 is missing extension for Route.delete. This includes it.
inline fun <reified T : Any> Route.delete(noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit): Route {
    return location(T::class) {
        method(HttpMethod.Delete) {
            handle(body)
        }
    }
}

object Paths {
    /**
     * 
     * 
     * @param userId  
     */
    @Location("/user/{userId}") class getUser(val userId: kotlin.Int)

}
```

##### Command line used for generation

```bash
openapi-codegen generate -g kotlin-server -i path/to/swagger.yaml -Dmodels -Dapis -DsupportingFiles=Paths.kt
```

##### Steps to reproduce

Generate the code using the config mentioned above.

##### Suggest a fix/enhancement

The template `Paths.kt.mustache` contains a check for bodyAllowed:

```mustache
object Paths {
{{#apis}}
{{#operations}}
    {{#operation}}
    {{^bodyAllowed}}
    /**
     * {{summary}}
     * {{#unescapedNotes}}{{.}}{{/unescapedNotes}}
     {{#allParams}}* @param {{paramName}} {{description}} {{^required}}(optional{{#defaultValue}}, default to {{{.}}}{{/defaultValue}}){{/required}}
     {{/allParams}}*/
    @Location("{{path}}") class {{operationId}}({{#allParams}}val {{paramName}}: {{{dataType}}}{{#hasMore}}, {{/hasMore}}{{/allParams}})

    {{/bodyAllowed}}
    {{/operation}}
{{/operations}}
{{/apis}}
}
```
Blindly removing it results in a NullPointerException.

### LINKED ISSUE #5640: [BUG][KOTLIN] kotlin-server generator bug that skips imports for enum classes in Paths.kt file
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description
The full sample project is published in https://github.com/svok/openapi-test
The bug concerns kotlin-server (KTOR) generator.
The main bug consequence is a lucking import declaration for the enum class in the Paths.kt file that causes `Unresolved reference: SomeStatus` error message.

##### openapi-generator version

4.2.3

##### OpenAPI declaration file content or url

Full declaration is over there: https://github.com/svok/openapi-test/blob/master/spec.yaml

The code causing error is following:
```yaml
  /pet/findByStatus:
    get:
      tags:
        - pet
      parameters:
        - name: status
          in: query
          schema:
            type: array
            items:
              $ref: '#/components/schemas/SomeStatus'
components:
  schemas:

    SomeStatus:
      type: string
      default: available
      enum:
        - available
        - pending
        - sold
```

##### Command line used for generation

The complete set of the parameters can be foun in the `generateKotlinModels` task of the https://github.com/svok/openapi-test/blob/master/build.gradle.kts file.

##### Steps to reproduce

To reproduce the error you need to start:
```bash
./gradlew generateKotlinModels
```
After that you can find in `Paths.kt` file:
```kotlin
    @KtorExperimentalLocationsAPI
    @Location("/pet/findByStatus") class findPetsByStatus(val status: kotlin.Array<SomeStatus>)
```
where `SomeStatus` is undeclared

The extra lines to be included into the file:
```kotlin
import org.example.rest.models.SomeStatus
```

*NOTE!* The same declaration in `kotlin/org/example/rest/models/Pet.kt` contains import for SomeStatus, but there it is not required.

##### Related issues/PRs

Not found

##### Suggest a fix

Right now I have no idea. The bug may relate to the following file: https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/kotlin-server/libraries/ktor/Paths.kt.mustache

### LINKED ISSUE #9087: Upgrade ktor server to the newest version
##### Description

Ktor server generator is outdated and it needs to be updated.

##### openapi-generator version

Latest (5.1.0)

### PR DESCRIPTION
**Pull request by saschpe**
Use Gradle 6.9 and Kotlin 1.4.32. Generate Paths for other HTTP verbs (#828) and fix imports (#5640). Use 'object' when no parameters are used. Introduce 'featureMetrics' to control metrics plugin usage. Remove HOCON configuration parsing. This is provided by `Application.environment.config already` and removes a dependency.

Resolves #9087, resolves #828, resolves #5640
Relates-To #5346

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jimschubert

### CODE REVIEW

**Comment by wing328** [comment]
@saschpe thanks for the PR. Is this change similar to https://github.com/OpenAPITools/openapi-generator/pull/9088?

Can you have a look at it to see if it meets your requirement as well?

**Comment by saschpe** [comment]
> @saschpe thanks for the PR. Is this change similar to #9088?
> 
> Can you have a look at it to see if it meets your requirement as well?

I didn't see this PR initially. It also improves quite a few things but it's not recent enough and misses a few things. Whenever that PR gets merged I'd be happy to rebase this one onto it and fix it.

**Comment by saschpe** [comment]
Once both PRs are merged (i.e. after review and potential reworks), I will change the serialization mechanism to be configurable in order to allow using `kotlinx.serialization`.

**Comment by wing328** [comment]
👌 I'll ping you here when #9088 gets merged.

**Comment by wing328** [comment]
I've merged https://github.com/OpenAPITools/openapi-generator/pull/9088 into `5.2.x` branch. Please kindly rebase your work on that and target the PR for `5.2.x` branch

**Comment by saschpe** [comment]
Ok

**Comment by saschpe** [comment]
Question: `master` or `5.2.x` ?

**Comment by wing328** [comment]
`5.2.x` (minor release - breaking changes with fallbacks) as that PR was merged into 5.2.x (instead of master)

Current master (patch release - no breaking changes) will be released soon.

**Comment by saschpe** [comment]
Rebased and re-targeted towards `5.2.x`

**Comment by wing328** [comment]
@saschpe thanks for the update.

cc @jimschubert (2017/09) ❤️, @dr4ke616 (2018/08) @karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10) @4brunu (2019/11) @yutaka0m (2020/03)

**Comment by rsinukov** [comment]
LGTM from Ktor side. 
@wing328 feel free to ping me in the future if you have changes related to the Ktor.

**Review by Manny27nyc — APPROVED** [review]

**Comment by saschpe** [comment]
The Travis-CI error looks unrelated but I lack the rights to re-build.

**Comment by 4brunu** [comment]
It looks unrelated.
To restart the CI you can close and reopen this PR.

**Comment by saschpe** [comment]
Uhm, wouldn't it be enough for a project maintainer to press re-build on Travis-CI rather than creating confusion here?

**Comment by 4brunu** [comment]
Sure, I was just trying to show you a quick way to restart CI.

**Comment by saschpe** [comment]
@jimschubert ping

**Comment by wing328** [comment]
Yes please rebase this one onto master.

I'll try to review tomorrow or this weekend.

**Comment by saschpe** [comment]
The Travis-CI error around `node_modules/@types/bluebird/index.d.ts(42,47): ` happens on other branches and merge requests as well.

**Comment by 4brunu** [comment]
Thanks, the remaining CI issues are not related to this PR 👍

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinServerCodegen.java [hunk 1]
```diff
@@ -268,17 +284,19 @@ public void processOpts() {
...
     public void postProcess() {
         System.out.println("################################################################################");
         System.out.println("# Thanks for using OpenAPI Generator.                                          #");
-        System.out.println("# Please consider donation to help us maintain this project \uD83D\uDE4F                 #");
+        System.out.println("# Please consider donation to help us maintain this project \uD83D\uDE4F                  #");
         System.out.println("# https://opencollective.com/openapi_generator/donate                          #");
         System.out.println("#                                                                              #");
         System.out.println("# This generator's contributed by Jim Schubert (https://github.com/jimschubert)#");
-        System.out.println("# Please support his work directly via https://patreon.com/jimschubert \uD83D\uDE4F      #");
+        System.out.println("# Please support his work directly via https://patreon.com/jimschubert \uD83D\uDE4F       #");
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinServerCodegen.java:299]: Now I got the following instead with the extra space that you added:
```
################################################################################
# Thanks for using OpenAPI Generator.                                          #
# Please consider donation to help us maintain this project 🙏                  #
# https://opencollective.com/openapi_generator/donate                          #
#                                                                              #
# This generator's contributed by Jim Schubert (https://github.com/jimschubert)#
# Please support his work directly via https://patreon.com/jimschubert 🙏       #
################################################################################
```

**Comment by wing328** [comment]
I got the following when running `gradle test` with the output:
```
> Task :compileKotlin
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/AppMain.kt: (50, 22): Unresolved reference: HttpClient
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/AppMain.kt: (50, 33): Unresolved reference: Apache
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/Configuration.kt: (55, 29): Unresolved reference: HttpMethod
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/Configuration.kt: (56, 24): Unresolved reference: settings
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/Configuration.kt: (57, 28): Unresolved reference: settings
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (30, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (32, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (32, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (61, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (63, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (63, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (69, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (71, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (71, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (100, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (102, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (102, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (131, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (133, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (162, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (164, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (164, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (193, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (195, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (195, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (201, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (203, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (203, 55): Unresolved reference: OAuthAccessTokenResponse
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/StoreApi.kt: (33, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/StoreApi.kt: (35, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (29, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (31, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (37, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (39, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (45, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (47, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (53, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (55, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (85, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (87, 30): Unresolved reference: authentication
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (93, 5): Unresolved reference: authenticate
e: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (95, 30): Unresolved reference: authentication

> Task :compileKotlin FAILED

```

Command to generate code: 
```
java -jar modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -g kotlin-server -i modules/openapi-generator/src/test/resources/3_0/petstore.yaml -o /tmp/kotlin-server
```

Can you please take a look when you've time?

**Comment by saschpe** [comment]
I'm still fiddleling with the the various sample projects. Not exactly straightforward ;-)

Actually, the Kotlin needs far bigger changes because there is quite some unnecessary HTTP handling built on top of tor-client. Also, the wide variety of JSON serialization and http client options don't make this any simpler.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinServerCodegen.java (hunk 1 above)
Comment (saschpe) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinServerCodegen.java]: It looked broken on my machine before. I think it's due to the UTF-32 glyphs. I moved back to the original version.

**Comment by saschpe** [comment]
> I got the following when running `gradle test` with the output:
> [...] 
> Can you please take a look when you've time?

Should be fixed.

**Comment by saschpe** [comment]
Anything left to do here?

**Comment by saschpe** [comment]
ping

**Comment by wing328** [comment]
Tested again and it's now built successfully:
```
$ gradle build
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileKotlin
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (33, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (64, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (72, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (103, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (134, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (165, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (196, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/PetApi.kt: (204, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/StoreApi.kt: (36, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (32, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (40, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (48, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (56, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (88, 13): Variable 'principal' is never used
w: /private/tmp/kotlin-server/src/main/kotlin/org/openapitools/server/apis/UserApi.kt: (96, 13): Variable 'principal' is never used

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0.2/userguide/command_line_interface.html#sec:command_line_warnings

```

**Pull request merged**